File: backend/patrol_api/gps_validation.py
"""
GPS Validation Utilities for R.O.N.D.A.
Comprehensive GPS accuracy, speed, and outlier detection.
"""

import math
from datetime import datetime, timedelta, timezone
from typing import Optional, Dict, Any, List, Tuple
from dataclasses import dataclass


@dataclass
class GPSValidationResult:
    """Result of GPS point validation"""
    is_valid: bool
    rejected_reason: Optional[str] = None
    accuracy_score: float = 0.0  # 0-1 scale, higher is better
    

@dataclass
class GPSPoint:
    """Represents a GPS coordinate with metadata"""
    latitude: float
    longitude: float
    timestamp: datetime
    accuracy: Optional[float] = None  # in meters
    speed: Optional[float] = None  # in m/s
    altitude: Optional[float] = None
    

class GPSValidator:
    """
    Comprehensive GPS validation with multiple quality checks.
    """
    
    # Configuration constants
    MAX_ACCEPTABLE_ACCURACY = 50.0  # meters
    MAX_GPS_AGE_MINUTES = 5  # GPS data must be less than 5 minutes old
    MAX_REALISTIC_SPEED_MPS = 55.56  # 200 km/h in m/s
    MAX_SPEED_JUMP_MPS = 27.78  # 100 km/h speed jump threshold
    MIN_DISTANCE_FOR_SPEED_CALC = 5.0  # meters
    OUTLIER_DISTANCE_THRESHOLD = 500.0  # meters from expected position
    MAX_REASONABLE_ACCELERATION = 8.0  # m/s² (0-100 km/h in 3.5s)
    
    # Philippines bounds (from existing code)
    MIN_LAT = 4.0
    MAX_LAT = 21.0
    MIN_LON = 112.0
    MAX_LON = 131.0
    
    def __init__(self):
        self.recent_points: List[GPSPoint] = []
        self.max_history = 5  # Keep last 5 points for outlier detection
    
    def validate_gps_point(
        self, 
        point: GPSPoint,
        previous_point: Optional[GPSPoint] = None
    ) -> GPSValidationResult:
        """
        Perform comprehensive validation on a single GPS point.
        
        Checks performed:
        1. Geographic bounds (Philippines)
        2. GPS accuracy threshold
        3. GPS age (timestamp not too old)
        4. Speed reasonableness
        5. Distance jump detection (outliers)
        6. Acceleration validation
        """
        
        # Check 1: Geographic bounds
        if not self._is_within_bounds(point.latitude, point.longitude):
            return GPSValidationResult(
                is_valid=False,
                rejected_reason=f"Coordinates out of bounds: ({point.latitude}, {point.longitude})",
                accuracy_score=0.0
            )
        
        # Check 2: Accuracy threshold
        if point.accuracy is not None and point.accuracy > self.MAX_ACCEPTABLE_ACCURACY:
            return GPSValidationResult(
                is_valid=False,
                rejected_reason=f"Poor accuracy: {point.accuracy:.1f}m > {self.MAX_ACCEPTABLE_ACCURACY}m threshold",
                accuracy_score=self._calculate_accuracy_score(point.accuracy)
            )
        
        # Check 3: GPS age
        age_check = self._check_gps_age(point.timestamp)
        if not age_check.is_valid:
            return age_check
        
        # Check 4-6: Require previous point for speed/outlier checks
        if previous_point is not None:
            # Check 4: Speed validation
            speed_check = self._validate_speed(point, previous_point)
            if not speed_check.is_valid:
                return speed_check
            
            # Check 5: Outlier detection (distance jump)
            outlier_check = self._detect_outlier(point, previous_point)
            if not outlier_check.is_valid:
                return outlier_check
            
            # Check 6: Acceleration validation
            accel_check = self._validate_acceleration(point, previous_point)
            if not accel_check.is_valid:
                return accel_check
        
        # Calculate overall accuracy score
        accuracy_score = self._calculate_overall_score(point)
        
        return GPSValidationResult(
            is_valid=True,
            rejected_reason=None,
            accuracy_score=accuracy_score
        )
    
    def _is_within_bounds(self, lat: float, lon: float) -> bool:
        """Check if coordinates are within Philippines bounds"""
        return (self.MIN_LAT <= lat <= self.MAX_LAT and 
                self.MIN_LON <= lon <= self.MAX_LON)
    
    def _check_gps_age(self, timestamp: datetime) -> GPSValidationResult:
        """Validate GPS data is not too old"""
        now = datetime.now(timezone.utc)
        
        # Ensure timestamp is timezone-aware
        if timestamp.tzinfo is None:
            timestamp = timestamp.replace(tzinfo=timezone.utc)
        
        age = now - timestamp
        max_age = timedelta(minutes=self.MAX_GPS_AGE_MINUTES)
        
        if age > max_age:
            return GPSValidationResult(
                is_valid=False,
                rejected_reason=f"Stale GPS data: {age.total_seconds()/60:.1f} minutes old > {self.MAX_GPS_AGE_MINUTES} min threshold",
                accuracy_score=0.0
            )
        
        return GPSValidationResult(is_valid=True, accuracy_score=1.0)
    
    def _validate_speed(
        self, 
        current: GPSPoint, 
        previous: GPSPoint
    ) -> GPSValidationResult:
        """Validate speed is realistic and not jumping unrealistically"""
        
        # If device provides speed, validate it
        if current.speed is not None:
            if current.speed > self.MAX_REALISTIC_SPEED_MPS:
                return GPSValidationResult(
                    is_valid=False,
                    rejected_reason=f"Unrealistic speed: {current.speed:.1f} m/s ({current.speed*3.6:.0f} km/h)",
                    accuracy_score=0.0
                )
        
        # Calculate speed between points
        distance = self._haversine_distance(
            previous.latitude, previous.longitude,
            current.latitude, current.longitude
        )
        
        time_diff = (current.timestamp - previous.timestamp).total_seconds()
        if time_diff > 0:
            calculated_speed = distance / time_diff  # m/s
            
            if calculated_speed > self.MAX_REALISTIC_SPEED_MPS:
                return GPSValidationResult(
                    is_valid=False,
                    rejected_reason=f"Calculated speed unrealistic: {calculated_speed:.1f} m/s ({calculated_speed*3.6:.0f} km/h)",
                    accuracy_score=0.0
                )
            
            # Check for speed jumps if we have previous speed
            if previous.speed is not None and current.speed is not None:
                speed_diff = abs(current.speed - previous.speed)
                if speed_diff > self.MAX_SPEED_JUMP_MPS:
                    return GPSValidationResult(
                        is_valid=False,
                        rejected_reason=f"Speed jump too large: {speed_diff:.1f} m/s change",
                        accuracy_score=0.3
                    )
        
        return GPSValidationResult(is_valid=True, accuracy_score=1.0)
    
    def _detect_outlier(
        self, 
        current: GPSPoint, 
        previous: GPSPoint
    ) -> GPSValidationResult:
        """Detect if point is an outlier based on distance from previous"""
        
        distance = self._haversine_distance(
            previous.latitude, previous.longitude,
            current.latitude, current.longitude
        )
        
        # Check if distance jump is unrealistic
        if distance > self.OUTLIER_DISTANCE_THRESHOLD:
            return GPSValidationResult(
                is_valid=False,
                rejected_reason=f"Distance jump too large: {distance:.0f}m > {self.OUTLIER_DISTANCE_THRESHOLD}m threshold",
                accuracy_score=0.0
            )
        
        return GPSValidationResult(is_valid=True, accuracy_score=1.0)
    
    def _validate_acceleration(
        self, 
        current: GPSPoint, 
        previous: GPSPoint
    ) -> GPSValidationResult:
        """Validate acceleration is within reasonable limits"""
        
        if current.speed is None or previous.speed is None:
            return GPSValidationResult(is_valid=True, accuracy_score=1.0)
        
        time_diff = (current.timestamp - previous.timestamp).total_seconds()
        if time_diff <= 0:
            return GPSValidationResult(is_valid=True, accuracy_score=1.0)
        
        speed_diff = current.speed - previous.speed
        acceleration = speed_diff / time_diff  # m/s²
        
        if abs(acceleration) > self.MAX_REASONABLE_ACCELERATION:
            return GPSValidationResult(
                is_valid=False,
                rejected_reason=f"Unrealistic acceleration: {acceleration:.1f} m/s²",
                accuracy_score=0.2
            )
        
        return GPSValidationResult(is_valid=True, accuracy_score=1.0)
    
    def _calculate_accuracy_score(self, accuracy: float) -> float:
        """Calculate accuracy score (0-1) based on accuracy in meters"""
        if accuracy <= 5:  # Excellent accuracy
            return 1.0
        elif accuracy <= 10:  # Good accuracy
            return 0.9
        elif accuracy <= 20:  # Acceptable
            return 0.7
        elif accuracy <= 50:  # Poor but usable
            return 0.4
        else:  # Unusable
            return 0.0
    
    def _calculate_overall_score(self, point: GPSPoint) -> float:
        """Calculate overall quality score for a GPS point"""
        scores = []
        
        # Accuracy score
        if point.accuracy is not None:
            scores.append(self._calculate_accuracy_score(point.accuracy))
        else:
            scores.append(0.5)  # Neutral if unknown
        
        # Speed score (prefer points with speed data)
        if point.speed is not None:
            scores.append(1.0)
        else:
            scores.append(0.5)
        
        return sum(scores) / len(scores) if scores else 0.5
    
    @staticmethod
    def _haversine_distance(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
        """Calculate distance between two coordinates in meters"""
        R = 6371000  # Earth's radius in meters
        
        phi1 = math.radians(lat1)
        phi2 = math.radians(lat2)
        delta_phi = math.radians(lat2 - lat1)
        delta_lambda = math.radians(lon2 - lon1)
        
        a = (math.sin(delta_phi / 2) ** 2 + 
             math.cos(phi1) * math.cos(phi2) * math.sin(delta_lambda / 2) ** 2)
        c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
        
        return R * c
    
    def filter_gps_points(
        self, 
        points: List[Dict[str, Any]], 
        strict_mode: bool = False
    ) -> Tuple[List[Dict[str, Any]], List[Dict[str, Any]]]:
        """
        Filter a list of GPS points, returning valid and rejected lists.
        
        Args:
            points: List of GPS point dictionaries
            strict_mode: If True, reject borderline points
            
        Returns:
            Tuple of (valid_points, rejected_points_with_reasons)
        """
        valid_points = []
        rejected_points = []
        
        previous_point = None
        
        for point_data in points:
            try:
                point = GPSPoint(
                    latitude=float(point_data['latitude']),
                    longitude=float(point_data['longitude']),
                    timestamp=point_data['timestamp'] if isinstance(point_data['timestamp'], datetime) else datetime.fromisoformat(str(point_data['timestamp'])),
                    accuracy=float(point_data['accuracy']) if 'accuracy' in point_data else None,
                    speed=float(point_data['speed']) if 'speed' in point_data else None,
                    altitude=float(point_data['altitude']) if 'altitude' in point_data else None
                )
                
                result = self.validate_gps_point(point, previous_point)
                
                if result.is_valid:
                    if not strict_mode or result.accuracy_score >= 0.7:
                        valid_points.append(point_data)
                        previous_point = point  # Update for next iteration
                    else:
                        rejected_points.append({
                            **point_data,
                            'rejected_reason': f"Low quality score: {result.accuracy_score:.2f}"
                        })
                else:
                    rejected_points.append({
                        **point_data,
                        'rejected_reason': result.rejected_reason
                    })
                    
            except (KeyError, ValueError, TypeError) as e:
                rejected_points.append({
                    **point_data,
                    'rejected_reason': f"Parse error: {str(e)}"
                })
        
        return valid_points, rejected_points


# Global validator instance
gps_validator = GPSValidator()

File: PNP-Patrol-App/lib/gpsValidation.ts
/**
 * GPS Validation Utilities for R.O.N.D.A. Mobile App
 * Validates GPS accuracy, speed, age, and detects outliers before sending to backend.
 */

import * as Location from 'expo-location';

// Validation Constants (matching backend values)
const MAX_ACCEPTABLE_ACCURACY = 50; // meters
const MAX_GPS_AGE_MINUTES = 5;
const MAX_REALISTIC_SPEED_MPS = 55.56; // 200 km/h
const MAX_SPEED_JUMP_MPS = 27.78; // 100 km/h speed jump threshold
const OUTLIER_DISTANCE_THRESHOLD = 500; // meters
const MAX_REASONABLE_ACCELERATION = 8.0; // m/s²

// Philippines bounds
const MIN_LAT = 4.0;
const MAX_LAT = 21.0;
const MIN_LON = 112.0;
const MAX_LON = 131.0;

/**
 * Interface for GPS validation result
 */
export interface GPSValidationResult {
  isValid: boolean;
  rejectedReason?: string;
  accuracyScore: number;
  warnings: string[];
}

/**
 * Interface for GPS point data
 */
export interface GPSPoint {
  latitude: number;
  longitude: number;
  timestamp: Date;
  accuracy?: number | null;
  speed?: number | null;
  altitude?: number | null;
}

/**
 * Haversine distance calculation between two coordinates in meters
 */
function haversineDistance(
  lat1: number,
  lon1: number,
  lat2: number,
  lon2: number
): number {
  const R = 6371000; // Earth's radius in meters
  
  const phi1 = (lat1 * Math.PI) / 180;
  const phi2 = (lat2 * Math.PI) / 180;
  const deltaPhi = ((lat2 - lat1) * Math.PI) / 180;
  const deltaLambda = ((lon2 - lon1) * Math.PI) / 180;
  
  const a =
    Math.sin(deltaPhi / 2) * Math.sin(deltaPhi / 2) +
    Math.cos(phi1) *
      Math.cos(phi2) *
      Math.sin(deltaLambda / 2) *
      Math.sin(deltaLambda / 2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  
  return R * c;
}

/**
 * Check if coordinates are within Philippines bounds
 */
function isWithinBounds(lat: number, lon: number): boolean {
  return lat >= MIN_LAT && lat <= MAX_LAT && lon >= MIN_LON && lon <= MAX_LON;
}

/**
 * Calculate accuracy score (0-1 scale)
 */
function calculateAccuracyScore(accuracy: number): number {
  if (accuracy <= 5) return 1.0;
  if (accuracy <= 10) return 0.9;
  if (accuracy <= 20) return 0.7;
  if (accuracy <= 50) return 0.4;
  return 0.0;
}

/**
 * Check GPS age is acceptable
 */
function checkGPSAge(timestamp: Date): { valid: boolean; ageMinutes: number } {
  const now = new Date();
  const ageMs = now.getTime() - timestamp.getTime();
  const ageMinutes = ageMs / (1000 * 60);
  
  return {
    valid: ageMinutes <= MAX_GPS_AGE_MINUTES,
    ageMinutes,
  };
}

/**
 * Main GPS validation function
 * Performs comprehensive validation on a GPS point
 */
export function validateGPSPoint(
  point: GPSPoint,
  previousPoint?: GPSPoint
): GPSValidationResult {
  const warnings: string[] = [];
  
  // Check 1: Geographic bounds
  if (!isWithinBounds(point.latitude, point.longitude)) {
    return {
      isValid: false,
      rejectedReason: `Coordinates out of bounds: (${point.latitude.toFixed(6)}, ${point.longitude.toFixed(6)})`,
      accuracyScore: 0,
      warnings,
    };
  }
  
  // Check 2: Accuracy threshold
  if (point.accuracy !== null && point.accuracy !== undefined) {
    if (point.accuracy > MAX_ACCEPTABLE_ACCURACY) {
      return {
        isValid: false,
        rejectedReason: `Poor accuracy: ${point.accuracy.toFixed(1)}m > ${MAX_ACCEPTABLE_ACCURACY}m threshold`,
        accuracyScore: calculateAccuracyScore(point.accuracy),
        warnings,
      };
    }
    
    // Warning for borderline accuracy
    if (point.accuracy > 30) {
      warnings.push(`Borderline accuracy: ${point.accuracy.toFixed(1)}m`);
    }
  }
  
  // Check 3: GPS age
  const ageCheck = checkGPSAge(point.timestamp);
  if (!ageCheck.valid) {
    return {
      isValid: false,
      rejectedReason: `Stale GPS data: ${ageCheck.ageMinutes.toFixed(1)} minutes old`,
      accuracyScore: 0.5,
      warnings,
    };
  }
  
  // Checks 4-6: Require previous point
  if (previousPoint) {
    const distance = haversineDistance(
      previousPoint.latitude,
      previousPoint.longitude,
      point.latitude,
      point.longitude
    );
    
    const timeDiffSeconds =
      (point.timestamp.getTime() - previousPoint.timestamp.getTime()) / 1000;
    
    // Check 4: Outlier detection (distance jump)
    if (distance > OUTLIER_DISTANCE_THRESHOLD) {
      return {
        isValid: false,
        rejectedReason: `Distance jump too large: ${distance.toFixed(0)}m`,
        accuracyScore: 0,
        warnings,
      };
    }
    
    // Check 5: Speed validation
    if (timeDiffSeconds > 0) {
      const calculatedSpeed = distance / timeDiffSeconds; // m/s
      
      if (calculatedSpeed > MAX_REALISTIC_SPEED_MPS) {
        return {
          isValid: false,
          rejectedReason: `Unrealistic speed: ${calculatedSpeed.toFixed(1)} m/s (${(
            calculatedSpeed * 3.6
          ).toFixed(0)} km/h)`,
          accuracyScore: 0,
          warnings,
        };
      }
      
      // Check speed jumps if we have previous speed
      if (
        previousPoint.speed !== null &&
        previousPoint.speed !== undefined &&
        point.speed !== null &&
        point.speed !== undefined
      ) {
        const speedDiff = Math.abs(point.speed - previousPoint.speed);
        
        if (speedDiff > MAX_SPEED_JUMP_MPS) {
          warnings.push(`Large speed jump: ${speedDiff.toFixed(1)} m/s`);
        }
      }
      
      // Check 6: Acceleration validation
      if (
        previousPoint.speed !== null &&
        previousPoint.speed !== undefined &&
        point.speed !== null &&
        point.speed !== undefined
      ) {
        const acceleration = (point.speed - previousPoint.speed) / timeDiffSeconds;
        
        if (Math.abs(acceleration) > MAX_REASONABLE_ACCELERATION) {
          return {
            isValid: false,
            rejectedReason: `Unrealistic acceleration: ${acceleration.toFixed(1)} m/s²`,
            accuracyScore: 0.2,
            warnings,
          };
        }
      }
    }
    
    // Warning for large but not extreme jumps
    if (distance > 100) {
      warnings.push(`Large position change: ${distance.toFixed(0)}m`);
    }
  }
  
  // Check device-provided speed if available
  if (point.speed !== null && point.speed !== undefined) {
    if (point.speed > MAX_REALISTIC_SPEED_MPS) {
      return {
        isValid: false,
        rejectedReason: `Device reports unrealistic speed: ${point.speed.toFixed(1)} m/s`,
        accuracyScore: 0,
        warnings,
      };
    }
  }
  
  // Calculate overall accuracy score
  let accuracyScore = 0.5; // Base score
  
  if (point.accuracy !== null && point.accuracy !== undefined) {
    accuracyScore = calculateAccuracyScore(point.accuracy);
  }
  
  // Boost score for having speed data
  if (point.speed !== null && point.speed !== undefined) {
    accuracyScore = Math.min(1.0, accuracyScore + 0.1);
  }
  
  return {
    isValid: true,
    accuracyScore,
    warnings,
  };
}

/**
 * Convert Expo Location object to GPSPoint
 */
export function locationToGPSPoint(location: Location.LocationObject): GPSPoint {
  return {
    latitude: location.coords.latitude,
    longitude: location.coords.longitude,
    timestamp: new Date(location.timestamp),
    accuracy: location.coords.accuracy,
    speed: location.coords.speed,
    altitude: location.coords.altitude,
  };
}

/**
 * Pre-validation check for Expo Location object
 * Quick check before more detailed validation
 */
export function quickValidateLocation(
  location: Location.LocationObject
): { valid: boolean; reason?: string } {
  // Check if location object is valid
  if (!location || !location.coords) {
    return { valid: false, reason: 'Invalid location object' };
  }
  
  const { latitude, longitude, accuracy } = location.coords;
  
  // Check for null/undefined coordinates
  if (latitude === null || latitude === undefined || 
      longitude === null || longitude === undefined) {
    return { valid: false, reason: 'Missing coordinates' };
  }
  
  // Check for zero coordinates (common GPS error)
  if (latitude === 0 && longitude === 0) {
    return { valid: false, reason: 'Zero coordinates (invalid GPS)' };
  }
  
  // Quick accuracy check
  if (accuracy !== null && accuracy !== undefined && accuracy > 100) {
    return { valid: false, reason: `Very poor accuracy: ${accuracy.toFixed(1)}m` };
  }
  
  return { valid: true };
}

/**
 * GPS Validation Manager
 * Maintains history of recent points for context-aware validation
 */
export class GPSValidationManager {
  private recentPoints: GPSPoint[] = [];
  private maxHistory: number = 5;
  
  /**
   * Add a point to history
   */
  addPoint(point: GPSPoint): void {
    this.recentPoints.push(point);
    
    // Keep only recent points
    if (this.recentPoints.length > this.maxHistory) {
      this.recentPoints.shift();
    }
  }
  
  /**
   * Get the most recent valid point
   */
  getLastValidPoint(): GPSPoint | undefined {
    return this.recentPoints.length > 0 
      ? this.recentPoints[this.recentPoints.length - 1] 
      : undefined;
  }
  
  /**
   * Clear history
   */
  clear(): void {
    this.recentPoints = [];
  }
  
  /**
   * Validate a new point with context from history
   */
  validate(point: GPSPoint): GPSValidationResult {
    const previousPoint = this.getLastValidPoint();
    const result = validateGPSPoint(point, previousPoint);
    
    if (result.isValid) {
      this.addPoint(point);
    }
    
    return result;
  }
  
  /**
   * Get validation statistics
   */
  getStats(): {
    historySize: number;
    lastPointTime?: string;
    lastPointCoords?: { lat: number; lon: number };
  } {
    const lastPoint = this.getLastValidPoint();
    
    return {
      historySize: this.recentPoints.length,
      lastPointTime: lastPoint?.timestamp.toISOString(),
      lastPointCoords: lastPoint 
        ? { lat: lastPoint.latitude, lon: lastPoint.longitude }
        : undefined,
    };
  }
}

// Global validation manager instance
export const gpsValidationManager = new GPSValidationManager();

/**
 * Format validation result for display/logging
 */
export function formatValidationResult(
  result: GPSValidationResult
): string {
  if (result.isValid) {
    let msg = `✅ Valid GPS (score: ${(result.accuracyScore * 100).toFixed(0)}%)`;
    if (result.warnings.length > 0) {
      msg += ` [Warnings: ${result.warnings.join(', ')}]`;
    }
    return msg;
  } else {
    return `❌ Invalid GPS: ${result.rejectedReason}`;
  }
}

/**
 * Get validation summary for batch processing
 */
export function getValidationSummary(
  results: GPSValidationResult[]
): {
  total: number;
  valid: number;
  invalid: number;
  averageScore: number;
  commonRejections: string[];
} {
  const valid = results.filter((r) => r.isValid).length;
  const invalid = results.length - valid;
  
  const scores = results
    .filter((r) => r.isValid)
    .map((r) => r.accuracyScore);
  const averageScore =
    scores.length > 0 ? scores.reduce((a, b) => a + b, 0) / scores.length : 0;
  
  // Count rejection reasons
  const rejectionCounts: Record<string, number> = {};
  results
    .filter((r) => !r.isValid && r.rejectedReason)
    .forEach((r) => {
      const reason = r.rejectedReason || 'Unknown';
      rejectionCounts[reason] = (rejectionCounts[reason] || 0) + 1;
    });
  
  const commonRejections = Object.entries(rejectionCounts)
    .sort((a, b) => b[1] - a[1])
    .slice(0, 3)
    .map(([reason, count]) => `${reason} (${count}x)`);
  
  return {
    total: results.length,
    valid,
    invalid,
    averageScore,
    commonRejections,
  };
}
